helix-core/src/main/java/org/apache/helix/controller/dataproviders/ResourceControllerDataProvider.java

```diff
@@ -294,4 +304,32 @@ public void clearMonitoringRecords() {
     _missingTopStateMap.clear();
     _lastTopStateLocationMap.clear();
   }
+
+  /**
+   * If the list items are the same, return the stable partition list cache. Otherwise, refresh
+   * the cache with new list and return the new list.
+   * This is for a backward compatible workaround to fix https://github.com/apache/helix/issues/940.
+   *
+   * @param resourceName
+   * @param newPartitionSet
+   */
+  public List<String> getOrSetStablePartitionList(String resourceName,
```

Comment:
This method is confusing. Another option is to call this method when the cluster data cache refresh IdealStates, this can keep all logic within data cache. 
i.e,
split to two methods: 1) refreshPartitionList(), this is private.  and 2) getPartitionList(), this is public and used by the rebalancer (and probably in other stages too where a partition list is needed, so there will be consistent over all stages).

Reply:
Changed with some additional modification to ensure safety.